libs/workscheduler.py

```diff
@@ -0,0 +1,199 @@
+       Once it begins, it may chose to run the work on another thread however.
+     - The final 'done' callback can happen on any thread, but will also use
+       set_timeout to move the completion work and UI updating to the UI thread
+
+    This design has the following benefits:
+     - Most tracking state is only touched by the UI thread, thus reducing the
+       need for complex locks or queues to avoid race conditions.
+     - Most actual Python processing happens on the one thread (assuming work
+       offloaded is mostly I/O bound work).  Due to the GIL, this is usually
+       the most efficient exeuction model for non-blocked work.
+
+
+    ## Example use of queue_request
+
+        # Set some locals to capture in the worker functions provided
+        _view = self.view
+        _file = self.filename
+        _loc  = self.view.location
+
+        # Define a function to do the request and notify on completion
+        def get_signature_data(on_done):
+            cli.request_signature_help(_file, _loc, on_done)
+
+        # Define a function to handle the completion response
+        def do_display(signature_data):
+            popup_text = get_sig_popup(signature_data)
+            _view.show_popup(popup_text)
+
+        # Schedule the request
+        queue_request(get_signature_data, do_display)
+    """
+
+    def __init__(self):
+        self.lock = threading.Lock()
+
+        # Set to the callback to be executed on the next schedule execution
+        self.next_job = None
+        # Set to the time the last job started execution
+        self.last_time = 0
+        # Set to the amount of time the last job took to execute
+        self.last_cost = 0
+        # Set to True if a timer is already pending
+        self.timer_set = False
+        # Set to True if a job is currently executing
+        self.job_running = False
+        # Set to True if the outstanding work has been canceled
+        self.canceled = False
+
+    def queue_request(self, worker, handler):
+        log.debug('In queue_request for work scheduler')
+
+        # Use nested functions to close over the worker and handler parameters
+
+        def work_done(results):
+            """ Called when the scheduled work item is complete
+
+            This function does some bookkeeping before calling the completion
+            handler provided when the job was queued.
+            """
+            log.debug('In work_done for work scheduler')
+            end_time = time.time()
+            canceled = False
+            with self.lock:
+                self.last_cost = end_time - self.last_time
+                self.job_running = False
+                canceled = self.canceled
+            log.debug('Work took {0:d}ms'.format(int(self.last_cost * 1000)))
+            if not canceled:
+                # Post the response to the handler on the main thread
+                sublime.set_timeout(lambda: handler(results), 0)
```

Comment:
See that comment on the line above it? :-)  That doesn't happen if you call it directly.